File: examples/dcps/ContentFilteredTopic/Java/src/ContentFilteredTopicDataPublisher.java
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to 2011 PrismTech 
 *   Limited and its licensees. All rights reserved. See file:
 *
 *                     $OSPL_HOME/LICENSE 
 *
 *   for full copyright notice and license terms. 
 *
 */

import java.util.Random;

import DDS.DataWriter;
import ContentFilteredTopicData.Stock;
import ContentFilteredTopicData.StockDataWriter;
import ContentFilteredTopicData.StockDataWriterHelper;
import ContentFilteredTopicData.StockTypeSupport;

public class ContentFilteredTopicDataPublisher {

	public static void main(String[] args) {
		DDSEntityManager mgr = new DDSEntityManager();
		String partitionName = "ContentFilteredTopic example";

		// create Domain Participant
		mgr.createParticipant(partitionName);

		// create Type
		StockTypeSupport stkTS = new StockTypeSupport();
		mgr.registerType(stkTS);

		// create Topic
		mgr.createTopic("StockTrackerExclusive");

		// create Publisher
		mgr.createPublisher();

		// create DataWriter
		mgr.createWriter();

		// Publish Events

		DataWriter dwriter = mgr.getWriter();
		StockDataWriter stockWriter = StockDataWriterHelper.narrow(dwriter);

		Stock msft = new Stock();
		msft.ticker = "MSFT";
		msft.price = 19.95f;

		Stock ge = new Stock();
		ge.ticker = "GE";
		ge.price = 10.00f;

		// register the two stock instances
		long msftHandle = stockWriter.register_instance(msft);
		long geHandle = stockWriter.register_instance(ge);

		// update data every second
		for (int x = 0; x < 20; x++) {
			msft.price = msft.price + (new Random()).nextFloat();
			stockWriter.write(msft, msftHandle);

			ge.price = ge.price + (new Random()).nextFloat();
			stockWriter.write(ge, geHandle);

			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}

		msft.price = -1.0f; // signal to terminate
		ge.price = -1.0f; // signal to terminate

		stockWriter.write(msft, msftHandle);
		stockWriter.write(ge, geHandle);
		try {
			Thread.sleep(1000);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}

		// clean up
		stockWriter.dispose(msft, msftHandle);
		stockWriter.dispose(ge, geHandle);
		stockWriter.unregister_instance(msft, msftHandle);
		stockWriter.unregister_instance(ge, geHandle);

		mgr.getPublisher().delete_datawriter(stockWriter);
		mgr.deletePublisher();
		mgr.deleteTopic();
		mgr.deleteParticipant();

	}
}

File: examples/dcps/ContentFilteredTopic/C++/src/ContentFilteredTopicDataPublisher.cpp

/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to 2011 PrismTech
 *   Limited and its licensees. All rights reserved. See file:
 *
 *                     $OSPL_HOME/LICENSE
 *
 *   for full copyright notice and license terms.
 *
 */

/************************************************************************
 * LOGICAL_NAME:    ContentFilteredTopicDataPublisher.cpp
 * FUNCTION:        OpenSplice Tutorial example code.
 * MODULE:          Tutorial for the C++ programming language.
 * DATE             September 2010.
 ************************************************************************
 *
 * This file contains the implementation for the 'ContentFilteredTopicDataPublisher' executable.
 *
 ***/
#include <string>
#include <sstream>
#include <iostream>
#include "DDSEntityManager.h"
#include "ccpp_ContentFilteredTopicData.h"
#include "os.h"


using namespace DDS;
using namespace ContentFilteredTopicData;

int main(int argc, char *argv[])
{
  os_time delay_100ms = { 0, 100000000 };
  DDSEntityManager *mgr = new DDSEntityManager();

  // create domain participant
  char partition_name[] = "ContentFilteredTopic example";
  mgr->createParticipant(partition_name);

  StockTypeSupport_var st = new StockTypeSupport();
  mgr->registerType(st.in());

  //create Topic
  char topic_name[] = "StockTrackerExclusive";
  mgr->createTopic(topic_name);

  //create Publisher
  mgr->createPublisher();

  // create DataWriter
  mgr->createWriter();
  DataWriter_ptr dWriter = mgr->getWriter();
  StockDataWriter_var ContentFilteredTopicDataWriter = StockDataWriter::_narrow(dWriter);

  Stock geQuote;
  Stock msftQuote;

  geQuote.ticker = CORBA::string_dup("GE");
  geQuote.price = 12.00f;
  msftQuote.ticker = CORBA::string_dup("MSFT");
  msftQuote.price = 25.00f;


  InstanceHandle_t geHandle = ContentFilteredTopicDataWriter->register_instance(geQuote);
  InstanceHandle_t msftHandle = ContentFilteredTopicDataWriter->register_instance(msftQuote);

  // Publish Events
  ReturnCode_t status;
  // update ContentFilteredTopicData price every second
  for (int x = 0; x < 20; x++)
  {
    geQuote.price = geQuote.price + 0.5;
    msftQuote.price = msftQuote.price + 1.5;
    printf("=== [ContentFilteredTopicDataPublisher] sends 2 stockQuotes : (GE, %.1f) (MSFT, %.1f)\n", geQuote.price, msftQuote.price);
    status = ContentFilteredTopicDataWriter->write(geQuote, geHandle);
    checkStatus(status, "StockDataWriter::write");
    status = ContentFilteredTopicDataWriter->write(msftQuote, msftHandle);
    checkStatus(status, "StockDataWriter::write");
    os_nanoSleep(delay_100ms);
  }
  // signal to terminate
  geQuote.price =  - 1;
  msftQuote.price =  - 1;
  ContentFilteredTopicDataWriter->write(geQuote, geHandle);
  ContentFilteredTopicDataWriter->write(msftQuote, msftHandle);
  cout << "Market Closed" << endl;

  /* Unregister the instances */
  ContentFilteredTopicDataWriter->unregister_instance(geQuote, geHandle);
  ContentFilteredTopicDataWriter->unregister_instance(msftQuote, msftHandle);

  /* Remove the DataWriters */
  mgr->deleteWriter(ContentFilteredTopicDataWriter.in ());

  /* Remove the Publisher. */
  mgr->deletePublisher();

  /* Remove the Topics. */
  mgr->deleteTopic();

  /* Remove Participant. */
  mgr->deleteParticipant();

  delete mgr;
  return 0;
}
